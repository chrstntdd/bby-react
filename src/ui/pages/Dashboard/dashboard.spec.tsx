import React from 'react';
import { shallow } from 'enzyme';

jest.useFakeTimers();

import ConnectedDashboard, { Dashboard } from './';

const requiredProps = {
  syncToDatabase: jest.fn(),
  printTable: jest.fn(),
  formatTable: jest.fn(),
  clearTable: jest.fn(),
  shuffleTable: jest.fn()
};

describe('<Dashboard/>', () => {
  it('should render without crashing', () => {
    shallow(<Dashboard {...requiredProps} />);
  });

  describe('componentDidMount method', () => {
    let render;
    beforeAll(() => {
      render = Dashboard.prototype.render;
      Dashboard.prototype.render = jest.fn();
    });

    afterAll(() => {
      Dashboard.prototype.render = render;
    });

    describe('when the component has products  ', () => {
      it('should call syncToDatabase method after 5 minutes and set the intervalId', () => {
        const wrapper = shallow(<Dashboard {...requiredProps} products={[{ test: 1 }]} />);

        jest.runOnlyPendingTimers();

        expect(wrapper.instance().intervalId).not.toBeNull();
        expect(wrapper.instance().intervalId).toEqual(expect.any(Number));
        expect(wrapper.instance().props.syncToDatabase).toHaveBeenCalledTimes(1);
      });
    });

    describe('when the component does not have any products', () => {
      it('should do nothing', () => {
        const wrapper = shallow(<Dashboard {...requiredProps} />);

        expect(wrapper.instance().intervalId).toBeNull();
      });
    });
  });

  describe('componentWillUnmount method', () => {
    let render;
    let componentDidMount;
    let clearInterval;
    beforeAll(() => {
      render = Dashboard.prototype.render;
      componentDidMount = Dashboard.prototype.componentDidMount;
      clearInterval = window.clearInterval;

      Dashboard.prototype.render = jest.fn();
      Dashboard.prototype.componentDidMount = jest.fn();
      window.clearInterval = jest.fn();
    });

    afterAll(() => {
      Dashboard.prototype.render = render;
      Dashboard.prototype.componentDidMount = componentDidMount;
      window.clearInterval = clearInterval;
    });

    afterEach(() => {
      window.clearInterval.mockClear();
    });

    describe('when there is a timer property set', () => {
      it('should call window.clearInterval', () => {
        const dashboard = new Dashboard({ ...requiredProps, products: [{ test: 1 }] });
        dashboard.intervalId = 1;

        dashboard.componentWillUnmount();

        expect(dashboard.intervalId).toBeNull();
        expect(window.clearInterval).toHaveBeenCalledTimes(1);
      });
    });

    describe('when there is **NOT** a timer property set', () => {
      it('should do nothing', () => {
        const dashboard = new Dashboard({ ...requiredProps });

        dashboard.componentWillUnmount();

        expect(dashboard.intervalId).toBeNull();
        expect(window.clearInterval).toHaveBeenCalledTimes(0);
      });
    });
  });

  describe('render method', () => {
    let componentDidMount;
    beforeAll(() => {
      componentDidMount = Dashboard.prototype.componentDidMount;
      Dashboard.prototype.componentDidMount = jest.fn();
    });

    afterAll(() => {
      Dashboard.prototype.componentDidMount = componentDidMount;
    });
    it('should render the dashboard', () => {
      const wrapper = shallow(<Dashboard {...requiredProps} />);

      expect(wrapper).toMatchSnapshot();
    });
  });
});
