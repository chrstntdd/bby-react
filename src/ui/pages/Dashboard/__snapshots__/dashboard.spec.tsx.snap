// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard/> render method should render the dashboard 1`] = `
<div
  id="dashboard-wrapper"
>
  <DashboardHeader />
  <div
    id="main-content-area"
  >
    <Component
      clearTable={[MockFunction]}
      formatTable={[MockFunction]}
      printTable={[MockFunction]}
      shuffleTable={[MockFunction]}
      syncToDatabase={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": undefined,
        }
      }
    />
    <section
      id="main-table-area"
    >
      <Connect(SearchBar) />
      <withRouter(Connect(ProductTable)) />
    </section>
  </div>
</div>
`;
